assets/js/utils/useRecordingChatMessages.ts

```diff
@@ -11,54 +11,55 @@ export const useRecordingChatMessages = (): {
   const [chatMessages, setChatMessages] = useState<ChatMessage[]>([]);
   const [isChatLoaded, setIsChatLoaded] = useState(false);
   const chatMessagesStack = useRef<AwaitingMessage[]>([]);
+  const requestedPreviousMessages = useRef(false);
 
-  const addMessage = useCallback((offset: number) => {
-    const messagesToAdd: AwaitingMessage[] = [];
-    let lastOnStack = chatMessagesStack.current.pop();
-    while (lastOnStack && offset >= lastOnStack.offset) {
-      messagesToAdd.push(lastOnStack);
-      lastOnStack = chatMessagesStack.current.pop();
+  const removeFromMessages = useCallback((prev: ChatMessage[], offset: number) => {
+    const messagesToRemove: AwaitingMessage[] = [];
+
+    let message = prev.pop();
+    let content = message ? message.contents.pop() : undefined;
+    while (message && content && content.offset > offset) {
+      messagesToRemove.push({ ...content, name: message.name, email: message.email });
+      if (message.contents.length === 0) message = prev.pop();
+      content = content = message ? message.contents.pop() : undefined;
     }
-    if (lastOnStack) chatMessagesStack.current.push(lastOnStack);
-    setChatMessages((prev) => appendToMessages(prev, messagesToAdd, undefined));
+    if (content && message) message.contents.push(content);
+    if (message) prev.push(message);
 
-    if (messagesToAdd.length === 0) {
-      setChatMessages((prev) => {
-        const messagesToRemove: AwaitingMessage[] = [];
+    chatMessagesStack.current.push(...messagesToRemove);
+    return [...prev];
+  }, []);
```

Comment:
I want to make sure how it works. We have two stacks one with messages already displayed and the second with messages to display. When we are removing messages we simply take displayed messages from the stack and push them to the stack with messages not displayed.
It also looks very similar to `addMessage` function so maybe we could create a generic function for both use cases?

Reply:
It works exactly as you explained it. I'm not sure about creating generic function as the logic behind pushing and poping from the shown messages is quite different, but I will try to simplify it a bit more.